src/lib/peers/peer-connection/PeerConnection.ts

```diff
@@ -0,0 +1,239 @@
+import PeerInfo from '../PeerInfo';
+import { randomBytes } from 'crypto';
+import * as grpc from '@grpc/grpc-js';
+import PeerManager from '../PeerManager';
+import TurnClient from '../turn/TurnClient';
+import KeyManager from '../../keys/KeyManager';
+import { stringToProtobuf, protobufToString, firstPromiseFulfilled } from '../../utils';
+import { PeerClient } from '../../../../proto/compiled/Peer_grpc_pb';
+import { PeerMessage, SubServiceType } from '../../../../proto/compiled/Peer_pb';
+
+class PeerConnection {
+  private publicKey: string;
+  private keyManager: KeyManager;
+  private peerManager: PeerManager;
+
+  private peerClient: PeerClient;
+  private connected: boolean = false;
+  private credentials: grpc.ChannelCredentials;
+
+  constructor(publicKey: string, keyManager: KeyManager, peerManager: PeerManager) {
+    this.publicKey = publicKey;
+    this.keyManager = keyManager;
+    this.peerManager = peerManager;
+
+    const pkiInfo = keyManager.PKIInfo;
+    if (pkiInfo.caCert && pkiInfo.cert && pkiInfo.key) {
+      this.credentials = grpc.credentials.createSsl(pkiInfo.caCert, pkiInfo.key, pkiInfo.cert);
+    } else {
+      this.credentials = grpc.credentials.createInsecure();
+    }
+  }
+
+  private async connectDirectly(): Promise<PeerClient> {
+    // try to create a direct connection
+    if (this.getPeerInfo().connectedAddress) {
+      // direct connection attempt
+      const address = this.getPeerInfo().connectedAddress!;
+      const peerClient = new PeerClient(address.toString(), this.credentials);
```

Comment:
The relationship between `PeerConnection` and `PeerClient` is quite confusing. As stated before, separate side effects from your objects as much as possible (at least for setup & teardown side effects).

Reply:
Just for clarity here, `PeerClient` is the client that is auto-generated by `protoc`, i.e. the actual grpc client. `PeerConnection` is the class abstracted over the top that contains the logic to select which of the 3 channels to use for connection: direct connection/relayed connection/UDP hole punched connection. 

Perhaps we could rename `PeerConnection` to something more appropriate or even add it into   the `PeerManager` class but I think it is a good separation of concerns. The `PeerServer` is another abstraction over the grpc server stub. I will add a note to the domain modelling issue about renaming of these classes to something less confusing.